akka-stream/src/main/scala/akka/stream/impl/io/InputStreamPublisher.scala

```diff
@@ -96,23 +59,20 @@ private[akka] class InputStreamPublisher(is: InputStream, bytesReadPromise: Prom
     readBytes match {
       case -1 ⇒
         // had nothing to read into this chunk
-        eofReachedAtOffset = readBytes
-        log.debug("No more bytes available to read (got `-1` or `0` from `read`), marking final bytes of file @ " + eofReachedAtOffset)
+        log.debug("No more bytes available to read (got `-1` from `read`)")
+        onCompleteThenStop()
 
       case _ ⇒
         readBytesTotal += readBytes
-        if (readBytes == chunkSize) availableChunks :+= ByteString1C(arr)
-        else availableChunks :+= ByteString1C(arr).take(readBytes)
 
-      // valid read, continue
+        // emit immediately, as this is the only chance to do it before we might block again
+        onNext(ByteString1C(arr).take(readBytes))
```

Comment:
Let's see. With the shared array the worst case scenario is if the `chunkSize` is such that `is.read(..)` almost always reads fully, then the space wasted is the size of the shared array.

With no shared array wasted space is for every read of size `chunkSize - readBytes`. Which could be small but also does not have a limit.

It really looks like shared array is less wasteful.